pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,167 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclaim
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/api/equality"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const tagNodeClaim = corev1beta1.Group + "/nodeclaim"
+const tagName = "Name"
+
+var managedTags = []string{tagName, tagNodeClaim}
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+}
+
+func NewController(kubeClient client.Client, instanceProvider *instance.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*corev1beta1.NodeClaim](kubeClient, &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *corev1beta1.NodeClaim) (reconcile.Result, error) {
+	if !isNodeClaimTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	// Extract resource ID from node spec
+	// TODO: Should we retry here? If the ProviderID is malformed is there a case where it is later repaired?
+	resourceID, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if instance has already been tagged
+	isTagged, err := c.isInstanceTagged(ctx, resourceID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if len(lo.PickByValues(isTagged, []bool{false})) != 0 {
+		tags := getTags(isTagged, nodeClaim)
+		err = c.tagInstance(ctx, tags, resourceID)
+		if err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				return reconcile.Result{Requeue: true}, err
```

Comment:
```suggestion
				return reconcile.Result{}, err
```

No need to specify `Requeue: true` if you return an error. It does the same thing. Logic is here: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/internal/controller/controller.go#L317